hack/codecoverage/README.md

```diff
@@ -0,0 +1,34 @@
+# Go Coverage tool
```

Comment:
why is the folder called "hack"?

Reply:
> A coverage percentage is not very actionable for developers. Would it not be better to provide more actionable information in a PR for both reviewers and PR author? For example, annotation / information about functions added in a PR that are not hit by unit tests, that would be really useful to start with.

That is way harder, and there are existing tools like VSCode, and Go compiler which can do that. 

> The config file is going to go out-of-sync as soon as we add files, etc.

Yes, if we don't meet 70% coverage. 

> Should this feature live in an external GHA rather than in this repo?

As a dev, I would like to run the checks locally, and this gives that option

>Making this blocking on PRs is probably just going to lead to the PR author reducing the threshold, since it's hard to tell which lines / functions are not tested.

The goal is to figure out the code coverage without using external dependencies. The code coverage is provided by VSCode and Go compiler. This isn't the design of this tool. Only then the code coverage is met. High-quality tests are critical for having confidence.


Right now, we are flying blind without knowing what the coverage of the project is. Any time new code gets added/updated, the project or the PR reviewer isn't aware of the test coverage.

What would be your suggestion to address this concern? Something like codecov would also suffice. Do you think we should enable an online tool?

Reply:
> As a dev, I would like to run the checks locally, and this gives that option

any reason why running `go test` locally does not suffice? 

> What would be your suggestion to address this concern? Something like codecov would also suffice. Do you think we should enable an online tool?

If codecov does it and they don't require write permissions to the repo, that sounds like an easy win

Reply:
>If codecov does it and they don't require write permissions, that sounds like an easy win, no?

@bobcallaway Do you have thoughts on codecov or similar other online tools?

Reply:
One thing we can do is open an HTML report of code coverage per file using the native go tools. Generate the coverage profile with go test, and then run `go tool cover`. We can make this a non-gating CI that allows reviewers (who ultimately should decide what can/cant be testable) to see if newly added code was tested. 

The CI job can also add a comment decomenting coverage % in each pkg. If all of this can be done without new code to maintain, that would be great.

e.g.
```
go test -v -coverprofile cover.out ./YOUR_CODE_FOLDER/...
go tool cover -html=cover.out -o cover.html
open cover.html
```

Reply:
+1 on having nothing to maintain ourselves.

Reply:
How do you make sure there is high test coverage without failing CI? Leaving it up to the reviewers to decide isn't going to get high enough coverage. 

Also, the PR requires 70% coverage. If the Reviewer agrees that 70% isn't attainable, then the reviewer can ask the PR author to update the file with different coverage.

Reply:
Codecov does have the ability to print an easy-to-read analysis of what the proposed change in the PR did to overall coverage, but it does not enforce a minimum level. Unless reviewers are actively evaluating the coverage (from codecov, the tool in this PR, or something else) then I would argue the measurement is being treated as noise.

Reply:
We have had good coverage without imposing a threshold so far. We'd like to keep it that way. Going back and forth between the reviewer and author to meet an arbitrary threshold is not worth it. 

What you describe about `the reviewer can ask the PR author to update the file with different coverage` is going to become the default behavior if we set arbitrary threshold.

Reply:
> Codecov does have the ability to print an easy-to-read analysis of what the proposed change in the PR did to overall coverage, but it does not enforce a minimum level. Unless reviewers are actively evaluating the coverage (from codecov, the tool in this PR, or something else) then I would argue the measurement is being treated as noise.

that's exactly the problem.

Reply:
I wouldn't say the threshold is arbitrary, I think it's more about setting a floor and then continuing to work upwards?

Reply:
>  Imposing arbitrary threshold is not.

> Although there is no “ideal code coverage number,” at Google we offer the general guidelines of 60% as “acceptable”, 75% as “commendable” and 90% as “exemplary.” However we like to stay away from broad top-down mandates and encourage every team to select the value that makes sense for their business needs.


source for ☝️  https://testing.googleblog.com/2020/08/code-coverage-best-practices.html

I think this blog post summarizes well.

Reply:
`A high code coverage percentage does not guarantee high quality in the test coverage` 
`More important than the percentage of lines covered is human judgment over the actual lines of code (and behaviors)  that aren’t being covered` 

Qualitative feedback is nice. Imposing threshold does not help IMO. I've seen this in scorecard: when the coverage goes down in a PR, we just let the PR go thru because that's the best we can do. Over time we don't even look at it because it's just noise and not actionable.

Having a general view of code coverage is great. Having it in PR as a percentage number provides no benefits IMO.

Reply:
OK cool! I understand if you have strong opinions on this 👍. I will close this PR.

Reply:
I'm not sure I understand your argument @laurentsimon - `we just let the PR go thru because thats the best we can do` - why not ask the contributor to add unit test coverage for what they're proposing to not erode the baseline percentage?

Reply:
1. I think having a floor is ok. Maybe somewhere around 50% minimum (though the 60% Google number is also fine)? But not necessarily any targets.
2. We should strive to ask for tests on PRs when appropriate.

I can go either way on making this a blocking pre-submit, but I agree there are merits either way.

> I'm not sure I understand your argument @laurentsimon - `we just let the PR go thru because thats the best we can do` - why not ask the contributor to add unit test coverage for what they're proposing to not erode the baseline percentage?-

I think that the biggest problem would be for a change that doesn't necessarily merit unit-tests or where unit-tests are especially hard to write, it might make doing it on that particular PR overly expensive just to maintain an arbitrary coverage baseline. We also have e2e tests that might cover certain scenarios better than unit-tests.

A better rule might be that we need to hit a baseline coverage % before we do a release?

Reply:
> A better rule might be that we need to hit a baseline coverage % before we do a release?

This is a good thing to enforce on "pre-release" -- if we do this it would be good to have a report generated per PR because accumulating tests at the end is a lot harder than per-change. 

> why not ask the contributor to add unit test coverage for what they're proposing to not erode the baseline percentage?

I think the problem was that per-file doesn't give reviewer or author enough actionable guidance. I still like the HTML report generated to help the reviewer evaluate test coverage of any added lines.